presto-cli/src/main/java/com/facebook/presto/cli/QueryPreprocessor.java

```diff
@@ -0,0 +1,164 @@
+                System.out.flush();
+            }
+            Signal.handle(SIGINT, oldHandler);
+            Thread.interrupted(); // clear interrupt status
+        }
+    }
+
+    private static String preprocessQueryInternal(Optional<String> catalog, Optional<String> schema, String query, List<String> preprocessorCommand, Duration timeout)
+            throws QueryPreprocessorException
+    {
+        // execute the process in a child thread so we can better handle interruption and timeouts
+        AtomicReference<Process> processReference = new AtomicReference<>();
+
+        FutureTask<String> task = new FutureTask<>(() -> {
+            String result;
+            int exitCode;
+            try {
+                ProcessBuilder processBuilder = new ProcessBuilder(preprocessorCommand);
+                processBuilder.redirectErrorStream(true);
+                processBuilder.environment().put(ENV_PRESTO_CATALOG, catalog.orElse(""));
+                processBuilder.environment().put(ENV_PRESTO_SCHEMA, schema.orElse(""));
+
+                Process process = processBuilder.start();
+                try {
+                    // write query to process standard out
+                    OutputStream outputStream = process.getOutputStream();
+                    outputStream.write(query.getBytes(UTF_8));
+                    process.getOutputStream().flush();
+                    process.getOutputStream().close();
+
+                    // read response
+                    result = CharStreams.toString(new InputStreamReader(process.getInputStream(), UTF_8));
+
+                    // wait for process to finish
+                    exitCode = process.waitFor();
+                }
+                finally {
+                    process.destroyForcibly();
+                }
+            }
+            catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                throw new QueryPreprocessorException("Interrupted while preprocessing query");
+            }
+            catch (Throwable e) {
+                throw new QueryPreprocessorException("Error preprocessing query: " + e.getMessage(), e);
+            }
+
+            // check we got a valid exit code
+            if (exitCode != 0) {
+                throw new QueryPreprocessorException("Query preprocessor failed:\n\n" + result);
+            }
+            return result;
+        });
+
+        Thread thread = new Thread(task);
+        thread.setName("Query preprocessor");
+        thread.start();
+
+        try {
+            return task.get(timeout.toMillis(), MILLISECONDS);
+        }
+        catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new QueryPreprocessorException("Interrupted while preprocessing query");
+        }
+        catch (ExecutionException e) {
+            Throwable cause = e.getCause();
+            propagateIfPossible(cause, QueryPreprocessorException.class);
```

Comment:
I believe other code like the pager, rethrow things like `Error`, which causes the CLI to exit (with stacktrace).  I think this is a good behavior in that case because the exception is truely unexpected and could be an OOM.